src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs

```diff
@@ -385,10 +383,299 @@ public static void EnsureSocketsAreInitialized()
+                    if (canonicalName == null && result->ai_canonname != IntPtr.Zero)
+                        canonicalName = Marshal.PtrToStringUni(result->ai_canonname);
+
+                    IPAddress ipAddress = null;
+
+                    if (result->ai_family == AddressFamily.InterNetwork)
+                    {
+                        ipAddress = CreateIPv4Address(result->ai_addr, result->ai_addrlen);
+                    }
+                    else if (SocketProtocolSupportPal.OSSupportsIPv6 && result->ai_family == AddressFamily.InterNetworkV6)
+                    {
+                        ipAddress = CreateIPv6Address(result->ai_addr, result->ai_addrlen);
+                    }
+
+                    if (ipAddress != null)
+                        addresses.Add(ipAddress);
+
+                    result = result->ai_next;
+                }
+
+                if (canonicalName == null)
+                    canonicalName = state.HostAddress;
+
+                state.CompleteAsyncResult(new IPHostEntry
+                {
+                    HostName = canonicalName,
+                    Aliases = Array.Empty<string>(),
+                    AddressList = addresses.ToArray()
+                });
+            }
+            finally
+            {
+                state.Dispose();
+            }
+        }
+
+        private static unsafe IPAddress CreateIPv4Address(byte* socketAddress, int addressLength)
+        {
+            const int IPv4SocketAddressSize = 16;
+
+            if (addressLength != IPv4SocketAddressSize)
+                return null;
+
+            long address = (long)(
+                               (socketAddress[4] & 0x000000FF) |
+                               (socketAddress[5] << 8 & 0x0000FF00) |
+                               (socketAddress[6] << 16 & 0x00FF0000) |
+                               (socketAddress[7] << 24)
+                           ) & 0x00000000FFFFFFFF;
+
+            return new IPAddress(address);
+        }
+
+        private static unsafe IPAddress CreateIPv6Address(byte* socketAddress, int addressLength)
+        {
+            const int IPv6SocketAddressSize = 28;
+            const int IPv6AddressBytes = 16;
+
+            if (addressLength != IPv6SocketAddressSize)
+                return null;
+
+            byte[] address = new byte[IPv6AddressBytes];
+            for (int i = 0; i < address.Length; i++)
+            {
+                address[i] = socketAddress[i + 8];
+            }
+
+            long scope = (long)((socketAddress[27] << 24) +
+                                (socketAddress[26] << 16) +
+                                (socketAddress[25] << 8) +
+                                (socketAddress[24]));
+
+            return new IPAddress(address, scope);
+        }
+
+        #region GetAddrInfoAsync Helper Classes
+
+        private sealed unsafe class GetAddrInfoExState : CriticalFinalizerObject, IDisposable
+        {
+            private IntPtr m_context;
```

Comment:
Can this be done as a SafeHandle-derived type rather than CriticalFinalizerObject?

Reply:
There is no reference counting and this handle is not used in a pinvoke, so I didn't see the gain in using SafeHandle. I used CriticalFinalizerObject for the CER, but since they don't exist in .Net Core I can make it derive from Object instead.